rs/replicated_state/src/replicated_state.rs

```diff
@@ -957,6 +969,80 @@ impl ReplicatedState {
         (expired_callback_count, errors)
     }
 
+    /// Enforces the best-effort message limit by repeatedly shedding the largest
+    /// best-effort messages of the canisters with the largest best-effort memory
+    /// usage until the memory usage drops below the limit.
+    ///
+    /// Returns the number of messages and message bytes that were shed.
+    ///
+    /// Time complexity: `O(n * log(n))`.
+    pub fn enforce_best_effort_message_limit(&mut self, limit: NumBytes) -> (u64, NumBytes) {
+        const ZERO: NumBytes = NumBytes::new(0);
+
+        // Check if we need to do anything at all before constructing a priority queue.
+        let mut memory_usage = self
```

Comment:
If we'd go for my suggestion above we could also use the `best_effort_message_memory_taken` function here.